File: src/main/java/com/github/tanokun/tanorpg/player/Member.java
package com.github.tanokun.tanorpg.player;

import com.github.tanokun.tanorpg.TanoRPG;
import com.github.tanokun.tanorpg.event.tanorpg.TanoRpgLevelUpEvent;
import com.github.tanokun.tanorpg.game.PermissionMap;
import com.github.tanokun.tanorpg.player.quest.QuestMap;
import com.github.tanokun.tanorpg.player.skill.SkillClass;
import com.github.tanokun.tanorpg.player.skill.SkillMap;
import com.github.tanokun.tanorpg.player.status.PlayerStatusMap;
import com.github.tanokun.tanorpg.player.status.StatusType;
import com.github.tanokun.tanorpg.player.warppoint.WarpPointMap;
import com.github.tanokun.tanorpg.util.io.Config;
import org.bukkit.Bukkit;
import org.bukkit.configuration.file.FileConfiguration;

import java.io.File;
import java.util.UUID;

import static org.bukkit.Bukkit.getOfflinePlayer;
import static org.bukkit.Bukkit.getPlayer;

public class Member {
    private final UUID uuid;

    private SkillClass skillClass;

    private final PlayerStatusMap statusMap;

    private final EquipmentMap equipMap;

    private final SkillMap skillMap;

    private final WarpPointMap warpPointMap;

    private final QuestMap questMap;

    private final PermissionMap permissionMap;

    private int hasHP;

    private int hasMP;

    private long hasEXP = 0;

    private MemberLevelType hasLevel;

    private Attack attack;

    public Member(UUID uuid, SkillClass skillClass, PlayerStatusMap pb, EquipmentMap equip, SkillMap skillMap, WarpPointMap warpPointMap, QuestMap questMap, PermissionMap permissionMap, Attack attack){
        this.uuid = uuid;
        this.skillClass = skillClass;
        this.statusMap = pb;
        this.equipMap = equip;
        this.skillMap = skillMap;
        this.warpPointMap = warpPointMap;
        this.questMap = questMap;
        this.permissionMap = permissionMap;
        this.attack = attack;
        this.hasLevel = MemberLevelType.Lv_1;
        this.hasHP = pb.getPointAndStatus(StatusType.HP);
        this.hasMP = pb.getPointAndStatus(StatusType.MP);
    }

    public Member(UUID uuid, SkillClass skillClass, PlayerStatusMap pb, EquipmentMap equip, SkillMap skillMap, WarpPointMap warpPointMap, QuestMap questMap, PermissionMap permissionMap, Attack attack, int hasHP, int hasMP, long hasEXP, MemberLevelType level) {
        this.uuid = uuid;
        this.skillClass = skillClass;
        this.statusMap = pb;
        this.equipMap = equip;
        this.skillMap = skillMap;
        this.warpPointMap = warpPointMap;
        this.questMap = questMap;
        this.permissionMap = permissionMap;
        this.attack = attack;
        this.hasLevel = level;
        this.hasHP = hasHP;
        this.hasMP = hasMP;
        this.hasEXP = hasEXP;
    }

    public void setMoney(long money) {
        TanoRPG.getPlugin().getEcon().depositPlayer(Bukkit.getPlayer(uuid), getMoney() - money);
        TanoRPG.getPlugin().getSidebarManager().updateSidebar(Bukkit.getPlayer(uuid),this);
    }

    public void removeMoney(long money) {
        TanoRPG.getPlugin().getEcon().withdrawPlayer(Bukkit.getPlayer(uuid), money);
        TanoRPG.getPlugin().getSidebarManager().updateSidebar(Bukkit.getPlayer(uuid),this);
    }

    public void addMoney(long money) {
        TanoRPG.getPlugin().getEcon().depositPlayer(Bukkit.getPlayer(uuid), money);
        TanoRPG.getPlugin().getSidebarManager().updateSidebar(Bukkit.getPlayer(uuid),this);
    }

    public void setHasHP(int hasHP) {
        this.hasHP = hasHP;
        TanoRPG.getPlugin().getSidebarManager().updateSidebar(Bukkit.getPlayer(uuid), this);

    }

    public void setHasMP(int hasMP) {
        this.hasMP = hasMP;
        TanoRPG.getPlugin().getSidebarManager().updateSidebar(Bukkit.getPlayer(uuid), this);

    }

    public void setHasEXP(long hasEXP) {
        this.hasEXP = hasEXP;

        if (!hasLevel.hasNext()){
            this.hasEXP = 0;
            return;
        }

        for (int i = 0; this.hasEXP >= hasLevel.getMaxEXP(); i++){
            if (!hasLevel.hasNext()){
                this.hasEXP = 0;
                TanoRPG.getPlugin().getSidebarManager().updateSidebar(Bukkit.getPlayer(uuid), this);
                return;
            }
            this.hasEXP = this.hasEXP - hasLevel.getMaxEXP();
            hasLevel = (hasLevel.getNext());
            Bukkit.getPlayer(uuid).sendMessage(TanoRPG.PX + "§aレベルが §b" + hasLevel + "Lv §aになりました！");
            getStatusMap().setStatusPoint(getStatusMap().getStatusPoint() + 2);
            Bukkit.getPluginManager().callEvent(new TanoRpgLevelUpEvent(getPlayer(uuid), this));
        }
        TanoRPG.getPlugin().getSidebarManager().updateSidebar(Bukkit.getPlayer(uuid), this);
    }

    public void setHasLevel(MemberLevelType hasLevel) {
        this.hasLevel = hasLevel;
        TanoRPG.getPlugin().getSidebarManager().updateSidebar(Bukkit.getPlayer(uuid), this);

    }

    public void setAttack(Attack attack) {
        this.attack = attack;
    }

    public UUID getUuid() {
        return uuid;
    }

    public SkillClass getSkillClass() {
        return skillClass;
    }

    public PlayerStatusMap getStatusMap() {
        return statusMap;
    }

    public EquipmentMap getEquipMap() {
        return equipMap;
    }

    public SkillMap getSkillMap() {
        return skillMap;
    }

    public WarpPointMap getWarpPointMap() {
        return warpPointMap;
    }

    public QuestMap getQuestMap() {
        return questMap;
    }

    public PermissionMap getOpenPermissionMap() {
        return permissionMap;
    }

    public long getMoney() {
        return Math.round(TanoRPG.getPlugin().getEcon().getBalance(getOfflinePlayer(uuid)));
    }

    public int getHasHP() {
        return hasHP;
    }

    public int getHasMP() {
        return hasMP;
    }

    public long getHasEXP() {
        return hasEXP;
    }

    public MemberLevelType getHasLevel() {
        return hasLevel;
    }

    public Attack getAttack() {
        return attack;
    }

    public void addHasEXP(long HAS_EXP){
        setHasEXP(getHasEXP() + HAS_EXP);
    }

    public void saveData(){
        Config data = new Config("player_database" + File.separator + uuid.toString() + ".yml", TanoRPG.getPlugin());
        FileConfiguration c = data.getConfig();
        data.createExists();
        c.set("skillClass", skillClass.name());
        statusMap.save(data, "");
        equipMap.save(data, "");
        questMap.save(data, "");
        permissionMap.save(data, "");
        c.set("hasHP", hasHP);
        c.set("hasMP", hasMP);
        c.set("hasEXP", hasEXP);
        c.set("hasLevel", hasLevel.name());
        data.saveConfig();
    }
}

File: src/main/java/com/github/tanokun/tanorpg/player/Member.kt
package com.github.tanokun.tanorpg.player

import com.github.tanokun.tanorpg.TanoRPG
import com.github.tanokun.tanorpg.event.tanorpg.TanoRpgLevelUpEvent
import com.github.tanokun.tanorpg.event.tanorpg.TanoRpgPlayerGetItemEvent
import com.github.tanokun.tanorpg.game.entity.boss.BossEntity
import com.github.tanokun.tanorpg.player.menu.main.backpack.BackpackMenu
import com.github.tanokun.tanorpg.player.quests.QuestMap
import com.github.tanokun.tanorpg.player.skill.SkillClass
import com.github.tanokun.tanorpg.player.skill.SkillMap
import com.github.tanokun.tanorpg.player.status.PlayerStatusMap
import com.github.tanokun.tanorpg.player.status.StatusType
import com.github.tanokun.tanorpg.player.status.buff.BuffMap
import com.github.tanokun.tanorpg.player.warppoint.WarpPointMap
import com.github.tanokun.tanorpg.util.io.Config
import org.bukkit.Bukkit
import org.bukkit.inventory.ItemStack
import java.io.File
import java.util.*
import kotlin.math.roundToInt
import kotlin.math.roundToLong

data class Member(val uuid: UUID, val skillClass: SkillClass, val statusMap: PlayerStatusMap, val equip: EquipmentMap, val skillMap: SkillMap,
                  val warpPointMap: WarpPointMap, val questMap: QuestMap, val buffMap: BuffMap, val attack: Attack,
                  val backpackMenu: BackpackMenu, val chestMap: ChestMap, var hasHP: Int = statusMap.getPointAndStatus(StatusType.HP).roundToInt(),
                  var hasMP: Int = statusMap.getPointAndStatus(StatusType.MP).roundToInt(), var hasEXP: Long = 0,
                  var level: MemberLevelType = MemberLevelType.Lv_1, val runnable: MemberRunnable, var bossEntity: BossEntity? = null){

    val player = Bukkit.getPlayer(uuid)!!

    fun setMoney(money: Long) {
        TanoRPG.plugin.econ.depositPlayer(Bukkit.getPlayer(uuid), (getMoney() - money).toDouble())
        TanoRPG.plugin.sidebarManager.updateSidebar(Bukkit.getPlayer(uuid), this)
    }

    fun removeMoney(money: Long) {
        TanoRPG.plugin.econ.withdrawPlayer(Bukkit.getPlayer(uuid), money.toDouble())
        TanoRPG.plugin.sidebarManager.updateSidebar(Bukkit.getPlayer(uuid), this)
    }

    fun addMoney(money: Long) {
        TanoRPG.plugin.econ.depositPlayer(Bukkit.getPlayer(uuid), money.toDouble())
        TanoRPG.plugin.sidebarManager.updateSidebar(Bukkit.getPlayer(uuid), this)
    }

    fun getMoney(): Long {
        return TanoRPG.plugin.econ.getBalance(Bukkit.getOfflinePlayer(uuid)).roundToLong()
    }

    fun online() {
        runnable.setMember(this)
        runnable.runTaskTimer(TanoRPG.plugin, 0, 20)
        player.setPlayerListName("[無所属] " + "§b§l" + player.name + "<Lv. " + level.value + ">")
    }

    fun offline() {
        runnable.cancel()
    }

    fun addItem(item: ItemStack) {
        player.inventory.addItem(item)
        Bukkit.getPluginManager().callEvent(TanoRpgPlayerGetItemEvent(player, item))
    }

    fun setHP(hasHP: Int) {
        this.hasHP = hasHP
        if (this.hasHP >= statusMap.getStatus(StatusType.HP)) this.hasHP = statusMap.getStatus(StatusType.HP).roundToInt()
        TanoRPG.plugin.sidebarManager.updateSidebar(Bukkit.getPlayer(uuid), this)
    }

    fun setMP(hasMP: Int) {
        this.hasMP = hasMP
        if (this.hasMP >= statusMap.getStatus(StatusType.MP)) this.hasMP = statusMap.getStatus(StatusType.MP).roundToInt()
        TanoRPG.plugin.sidebarManager.updateSidebar(Bukkit.getPlayer(uuid), this)
    }

    fun addHasEXP(hasEXP: Long) {
        this.hasEXP += hasEXP
        if (!level.hasNext()) {
            this.hasEXP = 0
            return
        }
        var i = 0
        while (this.hasEXP >= level.maxEXP) {
            if (!level.hasNext()) {
                this.hasEXP = 0
                TanoRPG.plugin.sidebarManager.updateSidebar(Bukkit.getPlayer(uuid), this)
                return
            }
            this.hasEXP = this.hasEXP - level.maxEXP
            level = level.next
            player.sendMessage(TanoRPG.PX + "§aレベルが §b" + level + "Lv §aになりました！")
            statusMap.statusPoint = statusMap.statusPoint + 2
            Bukkit.getPluginManager().callEvent(TanoRpgLevelUpEvent(player, this))
            i++
        }
        TanoRPG.plugin.sidebarManager.updateSidebar(Bukkit.getPlayer(uuid), this)
        player.setPlayerListName("[無所属] " + "§b§l" + player.name + "<Lv. " + level.value + ">")
    }

    fun saveData() {
        val data = Config(TanoRPG.plugin, "player_database" + File.separator + uuid.toString() + ".yml")
        val c = data.config
        data.createConfig()
        c["skillClass"] = skillClass.name
        statusMap.save(data, "")
        equip.save(data, "")
        questMap.save(data, "")
        skillMap.save(data, "")
        buffMap.save(data, "")
        backpackMenu.save(data, "")
        chestMap.save(data, "")
        c["hasHP"] = hasHP
        c["hasMP"] = hasMP
        c["hasEXP"] = hasEXP
        c["hasLevel"] = level.name
        data.saveConfig()
    }
}
